executor/new_distsql.go

```diff
@@ -286,62 +289,173 @@ type IndexLookUpExecutor struct {
-	handleCol *expression.Column
-
-	// result returns one or more distsql.PartialResult.
-	result distsql.SelectResult
-
-	taskChan chan *lookupTableTask
-	tasksErr error
-	taskCurr *lookupTableTask
-
+	handleCol    *expression.Column
 	tableRequest *tipb.DAGRequest
 	// columns are only required by union scan.
 	columns  []*model.ColumnInfo
 	priority int
+	// All fields above is immutable.
+
+	indexHandler
+	tableHandler
 	finished chan struct{}
 }
 
+// indexHandler is used by IndexLookUpExecutor to maintain index lookup background goroutines.
+type indexHandler struct {
+	wg sync.WaitGroup
+}
+
+// openIndexHandler open the indexHandler, launch a background goroutine to fetch handles, send the results to workCh.
+func (e *IndexLookUpExecutor) openIndexHandler(kvRanges []kv.KeyRange, workCh chan<- *lookupTableTask, finished <-chan struct{}) error {
+	result, err := distsql.SelectDAG(e.ctx.GetClient(), e.ctx.GoCtx(), e.dagPB, kvRanges,
+		e.ctx.GetSessionVars().DistSQLScanConcurrency, e.keepOrder, e.desc, getIsolationLevel(e.ctx.GetSessionVars()), e.priority)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	result.Fetch(e.ctx.GoCtx())
+	ih := &e.indexHandler
+	ih.wg.Add(1)
+	go func() {
+		ctx, cancel := goctx.WithCancel(e.ctx.GoCtx())
+		ih.fetchHandles(e, result, workCh, ctx, finished)
+		cancel()
+		if err := result.Close(); err != nil {
+			log.Error("close SelectDAG result failed:", errors.ErrorStack(err))
+		}
+		close(workCh)
+		ih.wg.Done()
+	}()
+	return nil
+}
+
+// fetchHandles fetches a batch of handles from index data and builds the index lookup tasks.
+func (ih *indexHandler) fetchHandles(e *IndexLookUpExecutor, result distsql.SelectResult, workCh chan<- *lookupTableTask, ctx goctx.Context, finished <-chan struct{}) {
+	for {
+		handles, finish, err := extractHandlesFromIndexResult(result)
+		if err != nil {
+			workCh <- &lookupTableTask{
+				tasksErr: errors.Trace(err),
+			}
+			return
+		}
+		if finish {
+			return
+		}
+		tasks := e.buildTableTasks(handles)
+		for _, task := range tasks {
+			select {
+			case <-ctx.Done():
+				return
+			case <-finished:
+				return
+			case workCh <- task:
+			}
+		}
+	}
+}
+
+func (ih *indexHandler) closeIndexHandler() {
+	ih.wg.Wait()
+}
+
+// tableHandler is used by IndexLookUpExecutor to maintain table lookup background goroutines.
+type tableHandler struct {
+	taskChan chan *lookupTableTask
+	taskCurr *lookupTableTask
+}
+
+// openTableHandler open the tableHandler, launch some background worker goroutines which pick
+// tasks from workCh, execute the task and store the results in tableHandler's taskChan.
+func (e *IndexLookUpExecutor) openTableHandler(workCh <-chan *lookupTableTask, finished <-chan struct{}) {
+	th := &e.tableHandler
+	th.taskChan = make(chan *lookupTableTask, atomic.LoadInt32(&LookupTableTaskChannelSize))
+	lookupConcurrencyLimit := e.ctx.GetSessionVars().IndexLookupConcurrency
+	var wg sync.WaitGroup
```

Comment:
How about make `wg` as a member of tableHandler like indexHandler?

Reply:
This variable is not necessary to be the inner state of `tableHandler`, a local variable is enough.
`tableHandler` close doesn't wait on it.